### DISCUSSION TITLE
/articles/sharepoint-framework-reimagine-flexible-generator/

### DISCUSSION THREAD

**Discussion by giscus** [Voitanos/voitanos.io-comments#318] [Comments]
Explore a proposal for a flexible SPFx project generator that could revolutionize SharePoint Framework development, offering customization and efficiency.
https://www.andrewconnell.com/articles/sharepoint-framework-reimagine-flexible-generator/

**Comment by DonKirkham** [comment]
💯agree. I don't see in the article how this de-couples Typescript and React versions.  I think your proposal is great for pro devs by allowing the use of custom templates, as long as the default configuration is simple for devs that are just getting started. SPFx toolkit offers similar options for adding PnPjs, PnP-React-Controls, and PnP-property-controls during project creation, having the ability to define in your personal template any other libraries you normally use is awesome.

While the project generator idea is great, I am also on board with changing the structure and dependencies in the project itself, allows devs to use the latest versions of all libraries. 

Yeoman was a valid option when SPFx first started, but I agree that SPFx is now mature enough to move beyond the limitations that come with that toolchain. I will track you down for further discussion!

**Reply by andrewconnell** [reply]
@DonKirkham said
>  I don't see in the article how this de-couples Typescript and React versions.

React is unique because it's a SPO dependency. But with TypeScript, I don't have to use their tooling. Their problem with TypeScript is the underlying dependencies in the build toolchain. But this approach lets others create their own toolchain and not rely on stuff like Heft or other build tools that require specific versions of TypeScript.

Yeoman was on the verge of being outdated when SPFx was initially introduced in 2017... even Angular was moving away from it at the time with their own CLI to create & manage projects.

**Comment by ynot3363** [comment]
I would love this approach, however I feel like we would still be locked into whichever versions of React SP is loading onto the page because of how they have modified it to run both react 16 and react 17 web parts on a page. However, maybe an update could include them adding more version support.

**Reply by andrewconnell** [reply]
Agreed... nothing can really address that as it's dependent on what they do in SPO, but at least I can change the templates to use React hooks! 🤣

**Comment by mikezimm** [comment]
I also agree with @Don and @ynot.
Having both standard and advanced methods helps reduce barriers for new people or those who just do not want the extra complexity.

I also echo the comments about the React versions... I really like the idea of leveraging the latest tech, but also found that as I build more and more projects that leverage the same npm library, those dependencies enhance what I can do for the effort, yet make it harder to maintain.  Double edge sword of trying to re-use code as much as possible.

**Reply by andrewconnell** [reply]
@mikezimm said:
> Having both standard and advanced methods helps reduce barriers for new people or those who just do not want the extra complexity.

Yup... the idea would be to have the same default templates baked into the utility that Microsoft already ships in the Yeoman generator, so the net result would be identical to what developers have today. But, we could create our own templates instead of relying on another extension or tool.

**Comment by darklordstudios** [comment]
Oh boy! I submitted my feedback and will make a comment. Several years ago, I switched to VueJS and have not looked back. This is my preferred framework. I use Vite and Vue to build components that I load with a modified SPFx script editor web part. It's a large lift, but works for the most part. I created a web part that allows users to create nice color coded calendars that are rich and modern and they love it. I would totally like to see a way to improve this to natively render a Vue component to load widgets and pass things like @pnp/graph and @pnp/sp so that I can continue improving the components that I build. Vue is such a nice framework and this would be something I would be willing to help do if I could.
